toolchain/sem_ir/inst_kind.h

```diff
@@ -6,10 +6,91 @@
 #define CARBON_TOOLCHAIN_SEM_IR_INST_KIND_H_
 
 #include <cstdint>
+#include <optional>
 
 #include "common/enum_base.h"
+#include "toolchain/parse/node_ids.h"
+
 namespace Carbon::SemIR {
 
+// Forward-declared to avoid a cycle.
+struct TypeId;
+
+// The expression category of a sem_ir instruction. See /docs/design/values.md
+// for details.
+enum class ExprCategory : int8_t {
+  // This instruction does not correspond to an expression, and as such has no
+  // category.
+  NotExpr,
+  // The category of this instruction is not known due to an error.
+  Error,
+  // This instruction represents a pattern, not an expression.
+  Pattern,
+  // This instruction represents a value expression.
+  Value,
+  // This instruction represents a durable reference expression, that denotes an
+  // object that outlives the current full expression context.
+  DurableRef,
+  // This instruction represents an ephemeral reference expression, that denotes
+  // an object that does not outlive the current full expression context.
+  EphemeralRef,
+  // This instruction represents an initializing expression, that describes how
+  // to initialize an object.
+  Initializing,
+  // This instruction represents a syntactic combination of expressions that are
+  // permitted to have different expression categories. This is used for tuple
+  // and struct literals, where the subexpressions for different elements can
+  // have different categories.
+  Mixed,
+  Last = Mixed
+};
+
+// What kind of expression category an instruction kind produces. The expression
+// category in general may depend on the operands of the instruction, but we can
+// handle most cases based on the instruction kind alone.
+class InstExprCategory {
+ public:
+  // The kind of category for this instruction. This includes all values in
+  // `ExprCategory`, for the case where the category of the instruction only
+  // depedns on its kind, plus some additional kinds.
+  enum Kind : int8_t {
```

Comment:
Consider adding `[[clang::enum_extensibility(open)]]` to help explain this enum holds other values too?

Alternatively, consider just using a `uint8_t` as the storage for the `kind_`, then casting to either `ExprCategory` or `Kind` (maybe `InstDependentKind`?). It felt a little weird to conceptually say `Kind` contains all the values of `ExprCategory` but then the return type of `AsInstDependentKind` is also `Kind` when we want to exclude `ExprCategory` values.

Reply:
Switched to `int8_t` for storage, and moved the kind enum out of the class. I also renamed it from `InstDependentKind` to `ComputedExprCategory`, as I think that better describes what it's doing -- and I'm now avoiding the word "dependent" to avoid suggesting this has anything to do with template dependence.